<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ резюме из HeadHunter
   

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px

# Исследование структуры данных

1. Прочитайте данные с помощью библиотеки Pandas. Совет: перед чтением обратите внимание на разделитель внутри файла. 

In [2]:
hh_database = pd.read_csv('D:\DS\data\dst-3.0_16_1_hh_database.csv', sep = ';')

<>:1: SyntaxWarning: invalid escape sequence '\D'
<>:1: SyntaxWarning: invalid escape sequence '\D'
C:\Users\1\AppData\Local\Temp\ipykernel_2988\1632024950.py:1: SyntaxWarning: invalid escape sequence '\D'
  hh_database = pd.read_csv('D:\DS\data\dst-3.0_16_1_hh_database.csv', sep = ';')


2. Выведите несколько первых (последних) строк таблицы, чтобы убедиться в том, что ваши данные не повреждены. Ознакомьтесь с признаками и их структурой.

In [3]:
display(hh_database.head(3))
hh_database.tail(3)

,"Пол, возраст",ЗП,Ищет работу на должность:,"Город, переезд, командировки",Занятость,График,Опыт работы,Последнее/нынешнее место работы,Последняя/нынешняя должность,Образование и ВУЗ,Обновление резюме,Авто
0,"Мужчина , 39 лет , родился 27 ноября 1979",29000 руб.,Системный администратор,"Советск (Калининградская область) , не готов к...","частичная занятость, проектная работа, полная ...","гибкий график, полный день, сменный график, ва...",Опыт работы 16 лет 10 месяцев Август 2010 — п...,"МАОУ ""СОШ № 1 г.Немана""",Системный администратор,Неоконченное высшее образование 2000 Балтийск...,16.04.2019 15:59,Имеется собственный автомобиль
1,"Мужчина , 60 лет , родился 20 марта 1959",40000 руб.,Технический писатель,"Королев , не готов к переезду , готов к редким...","частичная занятость, проектная работа, полная ...","гибкий график, полный день, сменный график, уд...",Опыт работы 19 лет 5 месяцев Январь 2000 — по...,Временный трудовой коллектив,"Менеджер проекта, Аналитик, Технический писатель",Высшее образование 1981 Военно-космическая ак...,12.04.2019 08:42,Не указано
2,"Женщина , 36 лет , родилась 12 августа 1982",20000 руб.,Оператор,"Тверь , не готова к переезду , не готова к ком...",полная занятость,полный день,Опыт работы 10 лет 3 месяца Октябрь 2004 — Де...,ПАО Сбербанк,Кассир-операционист,Среднее специальное образование 2002 Профессио...,16.04.2019 08:35,Не указано


,"Пол, возраст",ЗП,Ищет работу на должность:,"Город, переезд, командировки",Занятость,График,Опыт работы,Последнее/нынешнее место работы,Последняя/нынешняя должность,Образование и ВУЗ,Обновление резюме,Авто
44741,"Женщина , 48 лет , родилась 26 декабря 1970",40000 руб.,"Аналитик данных, Математик","Челябинск , готова к переезду , готова к редки...",полная занятость,"полный день, удаленная работа",Опыт работы 21 год 5 месяцев Январь 1998 — по...,"ОАО «ЧМК», Исследовательско-Технологический Це...",Начальник группы аналитики,Высшее образование 2000 Южно-Уральский госуда...,09.04.2019 05:07,Не указано
44742,"Мужчина , 24 года , родился 6 октября 1994",20000 руб.,Контент-менеджер,"Тамбов , не готов к переезду , не готов к кома...","частичная занятость, полная занятость",удаленная работа,Опыт работы 3 года 10 месяцев Контент-менедже...,IQ-Maxima,Менеджер проектов,Высшее образование 2015 Тамбовский государств...,26.04.2019 14:25,Имеется собственный автомобиль
44743,"Мужчина , 38 лет , родился 25 апреля 1980",120000 руб.,Руководитель проекта,"Москва , не готов к переезду , не готов к кома...",полная занятость,полный день,Опыт работы 15 лет 10 месяцев Руководитель пр...,ПАО ГК ТНС энерго,Руководитель отдела технической поддержки,Высшее образование 1997 Южно-Российский госуд...,05.07.2018 20:15,Не указано


In [4]:
hh_database.shape

(44744, 12)

3. Выведите основную информацию о числе непустых значений в столбцах и их типах в таблице.

In [5]:
hh_database.info()

<class 'pandas.DataFrame'>
RangeIndex: 44744 entries, 0 to 44743
Data columns (total 12 columns):
 #   Column                           Non-Null Count  Dtype
---  ------                           --------------  -----
 0   Пол, возраст                     44744 non-null  str  
 1   ЗП                               44744 non-null  str  
 2   Ищет работу на должность:        44744 non-null  str  
 3   Город, переезд, командировки     44744 non-null  str  
 4   Занятость                        44744 non-null  str  
 5   График                           44744 non-null  str  
 6   Опыт работы                      44576 non-null  str  
 7   Последнее/нынешнее место работы  44743 non-null  str  
 8   Последняя/нынешняя должность     44742 non-null  str  
 9   Образование и ВУЗ                44744 non-null  str  
 10  Обновление резюме                44744 non-null  str  
 11  Авто                             44744 non-null  str  
dtypes: str(12)
memory usage: 4.1 MB


4. Обратите внимание на информацию о числе непустых значений.

In [6]:
hh_database.notnull().sum()

Пол, возраст                       44744
ЗП                                 44744
Ищет работу на должность:          44744
Город, переезд, командировки       44744
Занятость                          44744
График                             44744
Опыт работы                        44576
Последнее/нынешнее место работы    44743
Последняя/нынешняя должность       44742
Образование и ВУЗ                  44744
Обновление резюме                  44744
Авто                               44744
dtype: int64

5. Выведите основную статистическую информацию о столбцах.

In [7]:
hh_database.describe(include=['object'])

C:\Users\1\AppData\Local\Temp\ipykernel_2988\1846862403.py:1: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  hh_database.describe(include=['object'])


,"Пол, возраст",ЗП,Ищет работу на должность:,"Город, переезд, командировки",Занятость,График,Опыт работы,Последнее/нынешнее место работы,Последняя/нынешняя должность,Образование и ВУЗ,Обновление резюме,Авто
count,44744,44744,44744,44744,44744,44744,44576,44743,44742,44744,44744,44744
unique,16003,690,14929,10063,38,47,44413,30214,16927,40148,18838,2
top,"Мужчина , 32 года , родился 17 сентября 1986",50000 руб.,Системный администратор,"Москва , не готов к переезду , не готов к кома...",полная занятость,полный день,Опыт работы 10 лет 8 месяцев Апрель 2018 — по...,Индивидуальное предпринимательство / частная п...,Системный администратор,Высшее образование 2014 Московский технически...,07.05.2019 09:50,Не указано
freq,18,4064,3099,1261,30026,22727,3,935,2062,4,25,32268


# Преобразование данных

1. Начнем с простого - с признака **"Образование и ВУЗ"**. Его текущий формат это: **<Уровень образования год выпуска ВУЗ специальность...>**. Например:
* Высшее образование 2016 Московский авиационный институт (национальный исследовательский университет)...
* Неоконченное высшее образование 2000  Балтийская государственная академия рыбопромыслового флота…
Нас будет интересовать только уровень образования.

Создайте с помощью функции-преобразования новый признак **"Образование"**, который должен иметь 4 категории: "высшее", "неоконченное высшее", "среднее специальное" и "среднее".

Выполните преобразование, ответьте на контрольные вопросы и удалите признак "Образование и ВУЗ".

Совет: обратите внимание на структуру текста в столбце **"Образование и ВУЗ"**. Гарантируется, что текущий уровень образования соискателя всегда находится в первых 2ух слов и начинается с заглавной буквы. Воспользуйтесь этим.

*Совет: проверяйте полученные категории, например, с помощью метода unique()*


In [8]:
# функция для преобразования признака "Образование и ВУЗ"
def education(string):
    '''
    Данная функция используестя для преобразования 
    признака "Образование и ВУЗ". Выделяет уровень образования
    при условии, что он записан в первых двух словах.

    Args:
        string (str): На вход подается строка,
        в которой слова разделены пробеами
    Returns:
        str: Возвращает строку с уровнем образования
    '''
    
    lst = ['высшее', 'специальное']
    
    string = string.split(' ')
    string = string[0: 2] 
    
    if string[1] not in lst:
        string = string[0]

    else:
        string = string[0]+' '+string[1]
    
    return string.lower()

In [9]:
# Проверяем работу функции
print(education('Среднее специальное образование 2017 Башкирск'))
print(education('Неоконченное высшее образование 2016 Новосиби'))
print(education('Высшее образование 2011 Московский Государстве'))
print(education('Среднее образование 2011 Московский Государстве'))

среднее специальное
неоконченное высшее
высшее
среднее


In [10]:
# создаем новый признак "Образование" 
hh_database['Образование'] = hh_database['Образование и ВУЗ'].apply(education)

# удаляем признак "Образование и ВУЗ" 
hh_database = hh_database.drop('Образование и ВУЗ', axis=1)

2. Теперь нас интересует столбец **"Пол, возраст"**. Сейчас он представлен в формате **<Пол , возраст , дата рождения >**. Например:
* Мужчина , 39 лет , родился 27 ноября 1979 
* Женщина , 21 год , родилась 13 января 2000
Как вы понимаете, нам необходимо выделить каждый параметр в отдельный столбец.

Создайте два новых признака **"Пол"** и **"Возраст"**. При этом важно учесть:
* Признак пола должен иметь 2 уникальных строковых значения: 'М' - мужчина, 'Ж' - женщина. 
* Признак возраста должен быть представлен целыми числами.

Выполните преобразование, ответьте на контрольные вопросы и удалите признак **"Пол, возраст"** из таблицы.

*Совет: обратите внимание на структуру текста в столбце, в части на то, как разделены параметры пола, возраста и даты рождения между собой - символом ' , '. 
Гарантируется, что структура одинакова для всех строк в таблице. Вы можете воспользоваться этим.*


In [11]:
# разделяем признак 'Пол, возраст' по ' , ' и оставляем только первые 2 элемента
hh_database['Пол, возраст'] = \
    hh_database['Пол, возраст'].apply(lambda x: x.split(' , ')[0:2])


# в признак 'Пол' записываем первый элемент ('М' - мужчина, 'Ж' - женщина)
hh_database['Пол'] = \
    hh_database['Пол, возраст'].apply(lambda x: 'М' if x[0]=='Мужчина' else 'Ж')


# в признак 'Возраст' записываем второй элемент преобразованный в int
hh_database['Возраст'] = \
    hh_database['Пол, возраст'].apply(lambda x: int(x[1].split(' ')[1]))

# удаляем признак 'Пол, возраст'
hh_database = hh_database.drop('Пол, возраст', axis=1)

In [12]:
# доля женщин в %
female_share = round(hh_database[hh_database['Пол']=='Ж'].shape[0]/hh_database.shape[0]*100, 2)
print(f'Доля женщин составила {female_share}%') 

# средний возраст
average_age = round(hh_database['Возраст'].mean(), 1)
print(f'Средний возраст составил {average_age} года')

Доля женщин составила 19.07%
Средний возраст составил 32.2 года


3. Следующим этапом преобразуем признак **"Опыт работы"**. Его текущий формат - это: **<Опыт работы: n лет m месяцев, периоды работы в различных компаниях…>**. 

Из столбца нам необходимо выделить общий опыт работы соискателя в месяцах, новый признак назовем "Опыт работы (месяц)"

Для начала обсудим условия решения задачи:
* Во-первых, в данном признаке есть пропуски. Условимся, что если мы встречаем пропуск, оставляем его как есть (функция-преобразование возвращает NaN)
* Во-вторых, в данном признаке есть скрытые пропуски. Для некоторых соискателей в столбце стоит значения "Не указано". Их тоже обозначим как NaN (функция-преобразование возвращает NaN)
* В-третьих, нас не интересует информация, которая описывается после указания опыта работы (периоды работы в различных компаниях)
* В-четвертых, у нас есть проблема: опыт работы может быть представлен только в годах или только месяцах. Например, можно встретить следующие варианты:
    * Опыт работы 3 года 2 месяца…
    * Опыт работы 4 года…
    * Опыт работы 11 месяцев…
    * Учитывайте эту особенность в вашем коде

Учитывайте эту особенность в вашем коде

В результате преобразования у вас должен получиться столбец, содержащий информацию о том, сколько месяцев проработал соискатель.
Выполните преобразование, ответьте на контрольные вопросы и удалите столбец **"Опыт работы"** из таблицы.


In [13]:
# В признаке 'Опыт работы':
# если значение NaN, пропускаем, 
# в противном случае разделяем по пробелу, оставляем 3-5 элемент включительно 
hh_database['Опыт работы'] = hh_database['Опыт работы'].apply(lambda x: x if x!=x 
                                                      else x.split(' ')[2:6])


# создаем функцию для преобразования признака 'Опыт работы'
def working_exp_month(lst): 
    '''
    Функция преобразует опыт работы из текстового формата 
    в количество лет.
    
    На вход могут подаваться списки следующего формата: 
    NaN
    []
    ['3', 'года', '2', 'месяца',]
    ['11', 'месяцев'] 
    ['4', 'года']
    
    Args:
        lst (list of strings): список строк

    Returns:
        1. Ничего не делает, если NaN
        2. Возвращает NaN, если подается пустой список (скрытое NaN)
        3. Умножает значение лет на 12 и прибаляет значение месяцев
    '''
       
    if lst != lst: pass  
    elif lst == []: return np.nan
    else:
        if lst[3].startswith('мес'):
            return int(lst[0])*12 + int(lst[2])
        
        elif lst[1].startswith('мес'):
              return int(lst[0])  
          
        else:
            return int(lst[0])*12
        
              
# создаем новую переменную 'Опыт работы (месяц)'
hh_database['Опыт работы (месяц)'] = \
    hh_database['Опыт работы'].apply(working_exp_month)


# удаляем переменную 'Опыт работы'
hh_database = hh_database.drop('Опыт работы', axis=1)

In [14]:
working_exp = hh_database['Опыт работы (месяц)'].median()
print(f'Медианный опыт работы {round(working_exp)} месяцев')

Медианный опыт работы 100 месяцев


4. Хорошо идем! Следующий на очереди признак "Город, переезд, командировки". Информация в нем представлена в следующем виде: **<Город , (метро) , готовность к переезду (города для переезда) , готовность к командировкам>**. В скобках указаны необязательные параметры строки. Например, можно встретить следующие варианты:

* Москва , не готов к переезду , готов к командировкам
* Москва , м. Беломорская , не готов к переезду, не готов к командировкам
* Воронеж , готов к переезду (Сочи, Москва, Санкт-Петербург) , готов к командировкам

Создадим отдельные признаки **"Город"**, **"Готовность к переезду"**, **"Готовность к командировкам"**. При этом важно учесть:

* Признак **"Город"** должен содержать только 4 категории: "Москва", "Санкт-Петербург" и "город-миллионник" (их список ниже), остальные обозначьте как "другие".

    Список городов-миллионников:
    
   <code>million_cities = ['Новосибирск', 'Екатеринбург','Нижний Новгород','Казань', 'Челябинск','Омск', 'Самара', 'Ростов-на-Дону', 'Уфа', 'Красноярск', 'Пермь', 'Воронеж','Волгоград']
    </code>
    Инфорация о метро, рядом с которым проживает соискатель нас не интересует.
* Признак **"Готовность к переезду"** должен иметь два возможных варианта: True или False. Обратите внимание, что возможны несколько вариантов описания готовности к переезду в признаке "Город, переезд, командировки". Например:
    * … , готов к переезду , …
    * … , не готова к переезду , …
    * … , готова к переезду (Москва, Санкт-Петербург, Ростов-на-Дону)
    * … , хочу переехать (США) , …
    
    Нас интересует только сам факт возможности или желания переезда.
* Признак **"Готовность к командировкам"** должен иметь два возможных варианта: True или False. Обратите внимание, что возможны несколько вариантов описания готовности к командировкам в признаке "Город, переезд, командировки". Например:
    * … , готов к командировкам , … 
    * … , готова к редким командировкам , …
    * … , не готов к командировкам , …
    
    Нас интересует только сам факт готовности к командировке.
    
    Еще один важный факт: при выгрузки данных у некоторых соискателей "потерялась" информация о готовности к командировкам. Давайте по умолчанию будем считать, что такие соискатели не готовы к командировкам.
    
Выполните преобразования и удалите столбец **"Город, переезд, командировки"** из таблицы.

*Совет: обратите внимание на то, что структура текста может меняться в зависимости от указания ближайшего метро. Учите это, если будете использовать порядок слов в своей программе.*


In [15]:
# создаем функцию, которая будет удалять информацию о метро 
# из признака "Город, переезд, командировки"
def metro(string):
    '''
    Удаляет иинформацию о метро из признака 
    "Город, переезд, командировки"

    Args:
        string (str): на вход подается строка
        'Москва ,  м. Сокол , не готов к переезду , готов к редким командировкам'

    Returns:
        list: возвращает список
        'Москва', 'не готов к переезду', 'готов к редким командировкам
        
    '''
    string = string.split(' , ')
    if string[1].lstrip().startswith('м. '): 
        string.pop(1)
    return string


# удаляем информацию о метро
hh_database["Город, переезд, командировки"] = \
    hh_database["Город, переезд, командировки"].apply(metro)



# Записываем в признак 'Город' первый эдемент из признака "Город, переезд, командировки"
hh_database['Город'] = hh_database["Город, переезд, командировки"].apply(lambda x:
    x[0:3].pop(0))


# Создаем список городов-миллионников
million_cities = ['Новосибирск', 'Екатеринбург', 'Нижний Новгород', 
                  'Казань', 'Челябинск', 'Омск', 'Самара', 'Ростов-на-Дону', 
                  'Уфа', 'Красноярск', 'Пермь', 'Воронеж', 'Волгоград' ]


# Преобразуем признак 'Город' в соответсвии с условием
hh_database['Город'] = hh_database['Город'].apply(lambda x: 
    x if x == 'Москва' or x == 'Санкт-Петербург' 
    else 'город-миллионник' if x in million_cities
    else 'другие')



# создаем lambda-функцию ready_nonready, которая возвращает True, 
# если строрка начинается с 'готов' или "хочу", False в другом случае
ready_nonready = lambda x: True if x.startswith(('г','хочу')) else False


# Записываем в 'Готовность к переезду' второй элемент из 
# "Город, переезд, командировки" Заменяем на True/False
hh_database['Готовность к переезду'] = \
    hh_database["Город, переезд, командировки"].apply(
    lambda x: x[0:3].pop(1))
    
hh_database['Готовность к переезду'] = \
    hh_database['Готовность к переезду'].apply(ready_nonready)



# удаляем первые два элемента из "Город, переезд, командировки"
hh_database["Город, переезд, командировки"] = \
    hh_database["Город, переезд, командировки"].apply(
    lambda x: x[2:3])

# Если в "Город, переезд, командировки" нет информации о 
# готовности к командировкам (т.е. []), заменяем его на ['не готов']
hh_database["Город, переезд, командировки"] = \
    hh_database["Город, переезд, командировки"].apply(
    lambda x: ['не готов'] if x == [] else x)



# Записываем в 'Готовность к командировкам' первый элемент из 
# "Город, переезд, командировки" Заменяем на True/False
hh_database['Готовность к командировкам'] = \
    hh_database["Город, переезд, командировки"].apply(lambda x: x[0])
    
hh_database['Готовность к командировкам'] = \
    hh_database['Готовность к командировкам'].apply(ready_nonready)


# Удаляем признак "Город, переезд, командировки" 
hh_database = hh_database.drop("Город, переезд, командировки", axis=1)


In [16]:
# Сколько процентов соискателей живут в Санкт-Петербурге? 
# Ответ округлите до целого.
mask = hh_database['Город'] == 'Санкт-Петербург'
SPB_share = hh_database[mask].shape[0]/hh_database.shape[0]*100
print(f'Доля соискателей, проживающих в Санкт-Петербурге {round(SPB_share)}%')


# Сколько процентов соискателей готовы одновременно и к переездам, и к командировкам? 
# Ответ округлите до целого.
mask_1 = hh_database['Готовность к переезду'] == True
mask_2 = hh_database['Готовность к командировкам'] == True
mov_err_share = hh_database[mask_1 & mask_2].shape[0]/hh_database.shape[0]*100
print(f'Доля соискателей, готовых к переезду и командировкам {round(mov_err_share)}%')

Доля соискателей, проживающих в Санкт-Петербурге 11%
Доля соискателей, готовых к переезду и командировкам 32%


5. Рассмотрим поближе признаки **"Занятость"** и **"График"**. Сейчас признаки представляют собой набор категорий желаемой занятости (полная занятость, частичная занятость, проектная работа, волонтерство, стажировка) и желаемого графика работы (полный день, сменный график, гибкий график, удаленная работа, вахтовый метод).
На сайте hh.ru соискатель может указывать различные комбинации данных категорий, например:
* полная занятость, частичная занятость
* частичная занятость, проектная работа, волонтерство
* полный день, удаленная работа
* вахтовый метод, гибкий график, удаленная работа, полная занятость

Такой вариант признаков имеет множество различных комбинаций, а значит множество уникальных значений, что мешает анализу. Нужно это исправить!

Давайте создадим признаки-мигалки для каждой категории: если категория присутствует в списке желаемых соискателем, то в столбце на месте строки рассматриваемого соискателя ставится True, иначе - False.

Такой метод преобразования категориальных признаков называется One Hot Encoding и его схема представлена на рисунке ниже:
<img src=https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/ohe.jpg>
Выполните данное преобразование для признаков "Занятость" и "График", ответьте на контрольные вопросы, после чего удалите их из таблицы

In [17]:
# Разделяем Занятость по ', '
hh_database['Занятость'] = hh_database['Занятость'].apply(lambda x: x.split(', '))


# Создаем признаки-мигалки по каждой категории
hh_database['полная занятость'] = hh_database['Занятость'].apply(
    lambda x: True if 'полная занятость' in x else False)

hh_database['частичная занятость'] = hh_database['Занятость'].apply(
    lambda x: True if 'частичная занятость' in x else False)

hh_database['проектная работа'] = hh_database['Занятость'].apply(
    lambda x: True if 'проектная работа' in x else False)

hh_database['волонтерство'] = hh_database['Занятость'].apply(
    lambda x: True if 'волонтерство' in x else False)

hh_database['стажировка'] = hh_database['Занятость'].apply(
    lambda x: True if 'стажировка' in x else False)


# Разделяем График по ', '
hh_database['График'] = hh_database['График'].apply(lambda x: x.split(', '))


# Создаем признаки-мигалки по каждой категории
hh_database['полный день'] = hh_database['График'].apply(
    lambda x: True if 'полный день' in x else False)

hh_database['сменный график'] = hh_database['График'].apply(
    lambda x: True if 'сменный график' in x else False)

hh_database['гибкий график'] = hh_database['График'].apply(
    lambda x: True if 'гибкий график' in x else False)

hh_database['удаленная работа'] = hh_database['График'].apply(
lambda x: True if 'удаленная работа' in x else False)

hh_database['вахтовый метод'] = hh_database['График'].apply(
    lambda x: True if 'вахтовый метод' in x else False)

In [18]:
# Сколько людей ищут проектную работу и волонтерство 
# (в обоих столбцах стоит True)?
mask_1 = hh_database['проектная работа'] == True
mask_2 = hh_database['волонтерство'] == True
employment = hh_database[mask_1 & mask_2].shape[0]
print(f'Проектную работу и волонтерство одновременно ищут {employment} человек')


# Сколько людей хотят работать вахтовым методом и с гибким графиком 
# (в обоих столбцах стоит True)?
mask_1 = hh_database['вахтовый метод'] == True
mask_2 = hh_database['гибкий график'] == True
schedule = hh_database[mask_1 & mask_2].shape[0]
print(f'Работать вахтовым методом или с гибким графиком желают {schedule} человек')


# Удаляем признаки 'Занятость' и 'График'
hh_database = hh_database.drop(['Занятость', 'График'], axis=1)

Проектную работу и волонтерство одновременно ищут 436 человек
Работать вахтовым методом или с гибким графиком желают 2311 человек


6. (2 балла) Наконец, мы добрались до самого главного и самого важного - признака заработной платы **"ЗП"**. 
В чем наша беда? В том, что помимо желаемой заработной платы соискатель указывает валюту, в которой он бы хотел ее получать, например:
* 30000 руб.
* 50000 грн.
* 550 USD

Нам бы хотелось видеть заработную плату в единой валюте, например, в рублях. Возникает вопрос, а где взять курс валют по отношению к рублю?

На самом деле язык Python имеет в арсенале огромное количество возможностей получения данной информации, от обращения к API Центробанка, до использования специальных библиотек, например pycbrf. Однако, это не тема нашего проекта.

Поэтому мы пойдем в лоб: обратимся к специальным интернет-ресурсам для получения данных о курсе в виде текстовых файлов. Например, MDF.RU, данный ресурс позволяет удобно экспортировать данные о курсах различных валют и акций за указанные периоды в виде csv файлов. Мы уже сделали выгрузку курсов валют, которые встречаются в наших данных за период с 29.12.2017 по 05.12.2019. Скачать ее вы можете **на платформе**

Создайте новый DataFrame из полученного файла. В полученной таблице нас будут интересовать столбцы:
* "currency" - наименование валюты в ISO кодировке,
* "date" - дата, 
* "proportion" - пропорция, 
* "close" - цена закрытия (последний зафиксированный курс валюты на указанный день).


Перед вами таблица соответствия наименований иностранных валют в наших данных и их общепринятых сокращений, которые представлены в нашем файле с курсами валют. Пропорция - это число, за сколько единиц валюты указан курс в таблице с курсами. Например, для казахстанского тенге курс на 20.08.2019 составляет 17.197 руб. за 100 тенге, тогда итоговый курс равен - 17.197 / 100 = 0.17197 руб за 1 тенге.
Воспользуйтесь этой информацией в ваших преобразованиях.

<img src=https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/table.jpg>


Осталось только понять, откуда брать дату, по которой определяется курс? А вот же она - в признаке **"Обновление резюме"**, в нем содержится дата и время, когда соискатель выложил текущий вариант своего резюме. Нас интересует только дата, по ней бы и будем сопоставлять курсы валют.

Теперь у нас есть вся необходимая информация для того, чтобы создать признак "ЗП (руб)" - заработная плата в рублях.

После ответа на контрольные вопросы удалите исходный столбец заработной платы "ЗП" и все промежуточные столбцы, если вы их создавали.

Итак, давайте обсудим возможный алгоритм преобразования: 
1. Перевести признак "Обновление резюме" из таблицы с резюме в формат datetime и достать из него дату. В тот же формат привести признак "date" из таблицы с валютами.
2. Выделить из столбца "ЗП" сумму желаемой заработной платы и наименование валюты, в которой она исчисляется. Наименование валюты перевести в стандарт ISO согласно с таблицей выше.
3. Присоединить к таблице с резюме таблицу с курсами по столбцам с датой и названием валюты (подумайте, какой тип объединения надо выбрать, чтобы в таблице с резюме сохранились данные о заработной плате, изначально представленной в рублях). Значение close для рубля заполнить единицей 1 (курс рубля самого к себе)
4. Умножить сумму желаемой заработной платы на присоединенный курс валюты (close) и разделить на пропорцию (обратите внимание на пропуски после объединения в этих столбцах), результат занести в новый столбец "ЗП (руб)".


In [19]:
# Подгружаем данные и удаляем ненудные столбцы
exchange = pd.read_csv('data/ExchangeRates.csv', sep = ',')
exchange = exchange.drop(['time', 'per', 'vol'], axis=1)

In [20]:
# Преобразуем дату в date-time формат и оставляем только дату
exchange['date'] = pd.to_datetime(exchange['date'], 
                                       format='%d/%m/%y').dt.date

# создаем доплнительный признак с обменным курсом и удаляем ненужные столбцы
exchange['exch_rate'] = exchange['close']/exchange['proportion']
exchange = exchange.drop(['close', 'proportion'], axis=1)


# в датафрейме hh_database создаем переменную date и сохраняем в нее дату из 
# 'Обновление резюме', которую затем удаляем
hh_database['date'] = \
    pd.to_datetime(hh_database['Обновление резюме'], dayfirst=True).dt.date



# В признак 'currency' записываем валюту из 'ЗП',
# в самой переменной 'ЗП' оставляем только значение ЗП
hh_database['currency'] = hh_database['ЗП'].apply(lambda x: x.split(' ')[1])
hh_database['ЗП'] = hh_database['ЗП'].apply(lambda x: int(x.split(' ')[0]))


# создаем функцию, которая будет переводить русскоязычные 
# названия валют в международные и применяем ее к признаку hh_database['currency']
def curr_name_changer (x): 
    if x=='бел.руб.':
        return 'BYN'
    elif x=='грн.':
        return 'UAH'
    elif x=='сум':
        return 'UZS'
    else:
        return x
        
hh_database['currency'] = hh_database['currency'].apply(curr_name_changer)    


# Соединяем два датафрейма по признакам 'date' и 'currency'
hh_database = hh_database.merge(exchange, how='left', on=['date', 'currency'])


# В переменной 'exch_rate' заменяем NaN на единицу
hh_database['exch_rate'] = hh_database['exch_rate'].apply(lambda x: 1 if x!=x else x)


# В 'ЗП (руб)' записываем результат умножения признаков 'ЗП' и 'exch_rate' 
hh_database['ЗП (руб)'] = hh_database['ЗП'] * hh_database['exch_rate']


# удаляем ненужные признаки 
hh_database.drop(['ЗП', 'exch_rate', 
              'date', 'currency'], 
             axis=1, inplace=True)


In [21]:
# Чему равна желаемая медианная заработная плата соискателей (в рублях)?
median_salary = hh_database['ЗП (руб)'].median()
print(f'Медианная желаемая ЗП составляет {round(median_salary)} руб.')

Медианная желаемая ЗП составляет 59019 руб.


In [22]:
hh_database.info()

<class 'pandas.DataFrame'>
RangeIndex: 44744 entries, 0 to 44743
Data columns (total 23 columns):
 #   Column                           Non-Null Count  Dtype  
---  ------                           --------------  -----  
 0   Ищет работу на должность:        44744 non-null  str    
 1   Последнее/нынешнее место работы  44743 non-null  str    
 2   Последняя/нынешняя должность     44742 non-null  str    
 3   Обновление резюме                44744 non-null  str    
 4   Авто                             44744 non-null  str    
 5   Образование                      44744 non-null  str    
 6   Пол                              44744 non-null  str    
 7   Возраст                          44744 non-null  int64  
 8   Опыт работы (месяц)              44574 non-null  float64
 9   Город                            44744 non-null  str    
 10  Готовность к переезду            44744 non-null  bool   
 11  Готовность к командировкам       44744 non-null  bool   
 12  полная занятость             

# Исследование зависимостей в данных

1. Постройте распределение признака **"Возраст"**. Опишите распределение, отвечая на следующие вопросы: чему равна мода распределения, каковы предельные значения признака, в каком примерном интервале находится возраст большинства соискателей? Есть ли аномалии для признака возраста, какие значения вы бы причислили к их числу?
*Совет: постройте гистограмму и коробчатую диаграмму рядом.*

In [23]:
# Отображаем значение моды
mode_age = hh_database['Возраст'].mode()[0]
print(f'Модальное значение возраста составляет {mode_age}')

Модальное значение возраста составляет 30


In [24]:
# Cоздаем и отображаем боксплот по признаку "Возраст"
fig_age_1 = px.box(
    data_frame=hh_database,
    title='Распределение признака "Возраст" <br> (боксплот)',
    x='Возраст',
    height=400,
    width=600
)

fig_age_1.show()
fig_age_1.write_html('data/fig_age_1_boxplot.html')


# Создаем и отображаем гистограмму по признаку "Возраст", а также 
# добавляем на нее линию, отражающую модальное значение признака
fig_age_2 = px.histogram(
    data_frame=hh_database,
    title='Распределение признака "Возраст" <br>(гистограмма)',
    x='Возраст',
    height=400,
    width=600,
)

fig_age_2.update_layout(bargap=0.1, yaxis_title='Кол-во соискателей')
fig_age_2.add_vline(x=hh_database['Возраст'].mode()[0], line_dash='dash', 
                    line_color='black', line_width=2.3)

fig_age_2.show()
fig_age_2.write_html('data/fig_age_2_hist.html')

***Выводы***:
1. медианное значение возраста - *31* год. 25 квантиль - *27* лет, 75 квантиль - *36* лет. 
2. модальное значение возраста составляет *30* лет, что не сильно отличается от медианного значения.
3. возраст распределен от *14* до *100* лет.
4. явным выборосом являются значения *76-77* лет, *100* лет, а также на мой взгляд *14-15* лет
5. Возраст большинства соискателей находится в пределах от 14 (lower fence) до 49 (upper frence) лет

---

2. Постройте распределение признака **"Опыт работы (месяц)"**. Опишите данное распределение, отвечая на следующие вопросы: чему равна мода распределения, каковы предельные значения признака, в каком примерном интервале находится опыт работы большинства соискателей? Есть ли аномалии для признака опыта работы, какие значения вы бы причислили к их числу?
*Совет: постройте гистограмму и коробчатую диаграмму рядом.*

In [25]:
# Cоздаем боксплот по признаку "Опыт работы (месяц)"
fig_exp_1 = px.box(
    data_frame=hh_database,
    title='Распределение признака "Опыт работы (месяц)" <br>(боксплот)',
    x='Опыт работы (месяц)',
    height=400,
    width=600
)
fig_exp_1.show()
fig_exp_1.write_html('data/fig_exp_1_boxplot.html')

# Создаем гистограмму по признаку "Опыт работы (месяц)", а также 
# добавляем на нее линию, отражающую модальное значение признака
fig_exp_2 = px.histogram(
    data_frame=hh_database,
    title='Распределение признака "Опыт работы (месяц)" <br>(гистограмма)',
    x='Опыт работы (месяц)',
    height=400,
    width=800
)
fig_exp_2.update_layout(bargap=0.1, yaxis_title='Кол-во соискателей')
fig_exp_2.add_vline(x=hh_database['Опыт работы (месяц)'].mode()[0], 
                    line_dash='dash', 
                    line_color='black', line_width=2.3)

fig_exp_2.show()
fig_exp_2.write_html('data/fig_exp_2_hist.html')

In [26]:
max_exp = hh_database['Опыт работы (месяц)'].max()
print(f'Максимальный опыт работы составил {round(max_exp)} месяц')

mode_exp = hh_database['Опыт работы (месяц)'].mode()[0]
print(f'Модальный опыт работы составил {round(mode_exp)} месяц')

Максимальный опыт работы составил 1188 месяц
Модальный опыт работы составил 81 месяц


***Выводы***:
1. Модальное значение признака "Опыт работы (месяц)" составило 81 месяц.
2. Минимальное значение признака составило 1 месяц, а максимальное 1188 мес.
3. Усы ящика имют значения от 1 месяца до 299 месяцев. Квантиль 25 - 57 месяцы, 75 квантиль - 154 месяца. В этом интервале находится опыт работы большинства соискателей.
4. К аномалиям я бы отнес всех соискателей, у которых опыт работы больше 550 месяцев (больше 45,8 лет рабочего стажа) - по 1 соискателю на отдельные значения. 1 выброс находится на уровне 1188 месяцев (99 лет стажа - 100% ошибка данных)

-----------

3. Постройте распределение признака **"ЗП (руб)"**. Опишите данное распределение, отвечая на следующие вопросы: каковы предельные значения признака, в каком примерном интервале находится заработная плата большинства соискателей? Есть ли аномалии для признака возраста? Обратите внимание на гигантские размеры желаемой заработной платы.
*Совет: постройте гистограмму и коробчатую диаграмму рядом.*


In [27]:
# Cоздаем боксплот по признаку "ЗП (руб)"
fig_sallary_1 = px.box(
    data_frame=hh_database,
    title='Распределение признака "ЗП (руб)" <br>(боксплот)',
    x='ЗП (руб)',
    height=400,
    width=800
)
fig_sallary_1.show()
fig_sallary_1.write_html('data/fig_sallary_1_boxplot.html')

# Создаем гистограмму по признаку "ЗП (руб)", а также 
# добавляем на нее линию, отражающую модальное значение признака
fig_sallary_2 = px.histogram(
    data_frame=hh_database,
    title='Распределение признака "ЗП (руб)" <br>(гистограмма)',
    x='ЗП (руб)',
    height=400,
    width=800 
)
fig_sallary_2.update_layout(bargap=0.1, yaxis_title='Кол-во соискателей')
fig_sallary_2.add_vline(x=hh_database['ЗП (руб)'].median(), line_dash='dash', 
                    line_color='black', line_width=2.3)

fig_sallary_2.show()
fig_sallary_2.write_html('data/fig_sallary_2_hist.html')

In [28]:
mode_sal = hh_database['ЗП (руб)'].mode()[0]
print(f'Модальный уровень ЗП составил {round(mode_sal)} месяц')
big_salary = hh_database[hh_database['ЗП (руб)']>10**6].shape[0]
print(f'ЗП более 1 млн. хотели бы иметь {big_salary} соискателей')

Модальный уровень ЗП составил 50000 месяц
ЗП более 1 млн. хотели бы иметь 5 соискателей


***Выводы***:
1. Мода распределения равна 50.000 рублей
2. Предельные значения уровня ЗП от 1 рубля до 24304880 рублей
3. Заработная плата большинства соискателей находится в интервале q1-q3 (37.055 - 95.000) тысяч рублей. Медианное значение - 
59.019 рублей - несколько отличается от модального значения
4. Аномалиями признака ЗП я бы посчитал все случаи выше 1 млн рублей: 1.75 млн, 2.5 млн, 3 млн, 7.67 млн, 24.3 млн
5. Заработную плату выше 1 миллиона рублей хотели бы иметь 5 соискателей

---------

4. Постройте диаграмму, которая показывает зависимость **медианной** желаемой заработной платы (**"ЗП (руб)"**) от уровня образования (**"Образование"**). Используйте для диаграммы данные о резюме, где желаемая заработная плата меньше 1 млн рублей.
*Сделайте выводы по представленной диаграмме: для каких уровней образования наблюдаются наибольшие и наименьшие уровни желаемой заработной платы? Как вы считаете, важен ли признак уровня образования при прогнозировании заработной платы?*

In [29]:
# Выбираем признаками ["ЗП (руб)", "Образование"] в отдельный датафрейм
df_edu_salary = hh_database[["ЗП (руб)", "Образование"]]

# Выбираем соискателей с ожидаемой ЗП ниже 1 млн. руб.
mask = df_edu_salary["ЗП (руб)"] < 10**6
salary_from_education = df_edu_salary[mask]

# Группируем по признаку "Образование" и определяем медианную ЗП
salary_from_education = salary_from_education.groupby('Образование', as_index=False).median()

# Отображаем данные в виде столбчатой диаграммы
fig = px.bar(
    data_frame=salary_from_education,
    x='Образование',
    y='ЗП (руб)',
    height=600,
    width=400,
    title='Зависимость ожидаемой ЗП <br> от уровня образования'
)
fig.update_layout(xaxis_title='Образование', 
                           yaxis_title='Ожидаемая зарплата в руб.')
fig.show()
fig.write_html('data/fig_sallary_from_education_bar.html')

***Выводы***:
1. наибольшая ожидаемая зарплата у соискателей с высшим образованием, наименьшая ожидаемая ЗП у соискателей со средним и средне-специальным образованием
2. Уровень ожидаемой ЗП одинаков для соискателей со средним и средне-специальным образованием
3. Я считаю, что признак "образование" является одним из ключевых при прогнозировании уровня ЗП

-------

5. Постройте диаграмму, которая показывает распределение желаемой заработной платы (**"ЗП (руб)"**) в зависимости от города (**"Город"**). Используйте для диаграммы данные о резюме, где желая заработная плата меньше 1 млн рублей.
*Сделайте выводы по полученной диаграмме: как соотносятся медианные уровни желаемой заработной платы и их размах в городах? Как вы считаете, важен ли признак города при прогнозировании заработной платы?*

In [30]:
# Выбираем признаками ["ЗП (руб)", "Город"] в отдельный датафрейм
df_city_salary = hh_database[["ЗП (руб)", "Город"]]

# Выбираем соискателей с ожидаемой ЗП ниже 1 млн. руб.
mask = df_city_salary["ЗП (руб)"] < 10**6
df_city_salary = df_city_salary[mask]

# Группируем по признаку "Город" и определяем медианную ЗП
salary_from_city = df_city_salary.groupby('Город', as_index=False).median()


# Отображаем данные в виде барплота
fig = px.bar(
    data_frame=salary_from_city,
    x='Город',
    y='ЗП (руб)',
    height=800,
    width=500,
    title='Зависимость ожидаемой ЗП от города'
)
fig.update_layout(xaxis_title='Город', 
                            yaxis_title='Ожидаемая зарплата в руб.')
fig.show()
fig.write_html('data/fig_sallary_from_city_bar.html')

***Выводы***:
1. максимальный уровень ожидаемой зарплаты в Москве 85.000 руб и Санкт-Петербурге - 60.000 рублей.
2. в городах-миллионниках и остальных городах медианная ожидаемая зп находится на уровне 40.000 рублей
3. я считаю, что признак города очень важен при прогнозировании уровня ЗП

---------

6. Постройте **многоуровневую столбчатую диаграмму**, которая показывает зависимость медианной заработной платы (**"ЗП (руб)"**) от признаков **"Готовность к переезду"** и **"Готовность к командировкам"**. Проанализируйте график, сравнив уровень заработной платы в категориях.

In [31]:
# Группируем медианную ЗП по готовности к переезду и командировкам
move_trips_data = hh_database.groupby(["Готовность к переезду", 
                                   "Готовность к командировкам"], 
                                  as_index=False)["ЗП (руб)"].median()


# Заменяем True/False на Готов/Не готов 
move_trips_data['Готовность к переезду'] = \
    move_trips_data['Готовность к переезду'].apply(
    lambda x: 'Готов' if x else 'Не готов')

move_trips_data['Готовность к командировкам'] = \
    move_trips_data['Готовность к командировкам'].apply(
    lambda x: 'Готов' if x else 'Не готов')

# Меняем местами строки, чтобы готовые соискатели шли вначале
move_trips_data = move_trips_data.reindex([3, 2, 1, 0]).reset_index(drop=True)


# Отображаем данные в виде многоуровнегого барплота
move_trips_fig = px.bar(
    data_frame=move_trips_data,
    y='ЗП (руб)',
    x='Готовность к переезду',
    color='Готовность к командировкам',
    title='Зависимость медианной ЗП от<br>\
готовности к переезду и командировкам',
    height=800,
    width=500
)
move_trips_fig.update_layout(yaxis_title='Ожидаемая зарплата в руб.', 
                             legend_title='Готовность <br>к командировкам')
move_trips_fig.show()
move_trips_fig.write_html('data/fig_move_trips.html')

***Выводы***:
1. Соискатели, готовые к переезду, ожидают большую ЗП нежели не готовые
2. Соискатели, которые готовы и к переезду, и к командировкам требуют наибольшую ЗП.
3. Наименьшую ЗП требуют соискатели не готовые ни к переезду, ни к командировкам.
4. Если соискатели готовы к переезду, готовность к командировкам влияет не так значительно на ожидаемую ЗП (разница примерно 15.000 рублей)
5. Если они не готовы к переезду, разница в ожидаемой ЗП составляет 20k руб.

--------

7. Постройте сводную таблицу, иллюстрирующую зависимость **медианной** желаемой заработной платы от возраста (**"Возраст"**) и образования (**"Образование"**). На полученной сводной таблице постройте **тепловую карту**. Проанализируйте тепловую карту, сравнив показатели внутри групп.

In [32]:
# Группируем данные по 'Образование' и 'Возраст', рассчитываем медиану
edu_age_slry = hh_database.groupby(['Образование', 'Возраст'], 
                               as_index=False)['ЗП (руб)'].median()


# Преобразуем данные из edu_age_slry в pivot-таблицу
pivot = pd.pivot_table(
    data=edu_age_slry,
    index='Образование',
    columns='Возраст',
    values='ЗП (руб)'
)


# Меняем местами названия строк
pivot = pivot.reindex(['высшее', 'неоконченное высшее', 
                       'среднее специальное', 'среднее'])


# Добавляем в pivot-таблицу недостающие колонки 
# в диапозоне от 14 до 100 заполненные NaN
lst = list(range(14,101))
pivot = pivot.reindex(columns=lst)


# Отображаем тепловую карту 
edu_age_fig = px.imshow(
    pivot,
    color_continuous_scale=['white', 'darkred'],
    title='Зависимость ожидаемой медианной ЗП \
от возраста и уровня образования'
)

# Чтобы отображались все значения возраста
edu_age_fig.update_layout(
    xaxis ={'tickmode':'array',
            'tickvals':list(range(14,101))}
    )

edu_age_fig.show()
edu_age_fig.write_html('data/pivot_age_edu_sal.html')

***Выводы***:
1. На тепловой карте мы видим, что уровень ожидаемой ЗП зависит от уровня образования (чем больше одно, тем больше другое).
2. Ожидания от ЗП увеличивааются с возрастом, особенно в группе с высшим образованием.
3. При этом наиболее интенсивно ожидания увеличиваются к 40-летнему возрасту.
4. Важно отметить присутствие соискателей с высшим образованием в возрастных группах до 21 года,
однако их доля очень небольшая и, по-видимому, эти данные являются аномальными.
5. В группах со средним образованием ожидания ЗП с возрастом увеличиваются значительно менее интенсивно.
6. После 60-ти летнего возраста, соискатели преиущественно представлены людьми с высшим образованием.

--------------

8. Постройте **диаграмму рассеяния**, показывающую зависимость опыта работы (**"Опыт работы (месяц)"**) от возраста (**"Возраст"**). Опыт работы переведите из месяцев в года, чтобы признаки были в едином масштабе. Постройте на графике дополнительно прямую, проходящую через точки (0, 0) и (100, 100). Данная прямая соответствует значениям, когда опыт работы равен возрасту человека. Точки, лежащие на этой прямой и выше нее - аномалии в наших данных (опыт работы больше либо равен возрасту соискателя)

In [33]:
# Создаем скеттерплот, в котором опыт работы зависит от возраста
exp_age_fig = px.scatter(
    data_frame=hh_database,
    y=hh_database['Опыт работы (месяц)']/12,
    x='Возраст',
    title='Зависимость опыта работы <br>от возраста',

    height=800,
    width=800,
)


# Уменьшаем точки, добавляем название оси Y и линию-разделитель
exp_age_fig.update_traces(marker={'size':5})
exp_age_fig.update_layout(yaxis_title='Опыт работы (лет)')
exp_age_fig.add_shape(type='line',
                      x0=0, y0=0,
                      x1=100, y1=100,
                      line={'color':'darkred',
                            'dash':'dash'})
exp_age_fig.show()
exp_age_fig.write_html('data/scatter_exp_age.html')

***Выводы***:
1. наблюдается прямая зависимость опыта работы от возраста
2. на диаграмме четко заметны выброс справа - возраст 100 лет, опыт работы минимален. Так же порядка 10 выбросов сверху, где опыт работы либо превышает возраст соискателя, либо работать они начали официально в 4-7 лет

**Дополнительные баллы**

Для получения 2 дополнительных баллов по разведывательному анализу постройте еще два любых содержательных графика или диаграммы, которые помогут проиллюстрировать влияние признаков/взаимосвязь между признаками/распределения признаков. Приведите выводы по ним. Желательно, чтобы в анализе участвовали признаки, которые мы создавали ранее в разделе "Преобразование данных".


Построим диаграмму, которая показывает распределение желаемой заработной платы (**"ЗП (руб)"**) в зависимости от пола(**"Пол"**). Используем для диаграммы данные о резюме, где желая заработная плата меньше 1 млн рублей.

In [34]:
# Выбираем признаками ["ЗП (руб)", "Город"] в отдельный датафрейм
df_gender_salary = hh_database[["ЗП (руб)", "Пол"]]

# Выбираем соискателей с ожидаемой ЗП ниже 1 млн. руб.
mask = df_gender_salary["ЗП (руб)"] < 10**6
df_gender_salary = df_gender_salary[mask]

# Группируем по признаку "Город" и определяем медианную ЗП
salary_from_gender = df_gender_salary.groupby('Пол', as_index=False).median()


# Отображаем данные в виде барплота
fig = px.bar(
    data_frame=salary_from_gender,
    x='Пол',
    y='ЗП (руб)',
    height=800,
    width=500,
    title='Зависимость ожидаемой ЗП от Пола'
)
fig.update_layout(xaxis_title='Пол', 
                            yaxis_title='Ожидаемая зарплата в руб.')
fig.show()
fig.write_html('data/fig_sallary_from_gender_bar.html')

***Выводы***:
Медианная ожидаемая зарплата у женщин оставляет 48.000 руб., у мужчин - 60.000 руб.

# Очистка данных

1. Начнем с дубликатов в наших данных. Найдите **полные дубликаты** в таблице с резюме и удалите их. 

In [35]:
# Создаем маску для выделения дубликатов, а затем удаляем их
mask = hh_database.duplicated(subset=hh_database.columns)
print(f'Число полных дубликатов составляет {hh_database[mask].shape[0]}')
hh_database.drop_duplicates(inplace=True)

Число полных дубликатов составляет 158


2. Займемся пропусками. Выведите информацию **о числе пропусков** в столбцах. 

In [36]:
# Выводим данные об NaN во всех переменных
hh_database.info()

# Определяем сколько пропусков в 'Опыт работы (месяц)'
cols_with_nan = hh_database['Опыт работы (месяц)'].isnull()
nan_by_exp = hh_database[cols_with_nan].shape[0]
print()

print(f'В признаке "Опыт работы (месяц)" кол-во NaN составляет {nan_by_exp}')

<class 'pandas.DataFrame'>
Index: 44586 entries, 0 to 44743
Data columns (total 23 columns):
 #   Column                           Non-Null Count  Dtype  
---  ------                           --------------  -----  
 0   Ищет работу на должность:        44586 non-null  str    
 1   Последнее/нынешнее место работы  44585 non-null  str    
 2   Последняя/нынешняя должность     44584 non-null  str    
 3   Обновление резюме                44586 non-null  str    
 4   Авто                             44586 non-null  str    
 5   Образование                      44586 non-null  str    
 6   Пол                              44586 non-null  str    
 7   Возраст                          44586 non-null  int64  
 8   Опыт работы (месяц)              44418 non-null  float64
 9   Город                            44586 non-null  str    
 10  Готовность к переезду            44586 non-null  bool   
 11  Готовность к командировкам       44586 non-null  bool   
 12  полная занятость                 4

3. Итак, у нас есть пропуски в 3ех столбцах: **"Опыт работы (месяц)"**, **"Последнее/нынешнее место работы"**, **"Последняя/нынешняя должность"**. Поступим следующим образом: удалите строки, где есть пропуск в столбцах с местом работы и должностью. Пропуски в столбце с опытом работы заполните **медианным** значением.

In [37]:
# Удаляем строки с NaN в признаках 
# 'Последнее/нынешнее место работы' и 'Последняя/нынешняя должность'
hh_database = hh_database.dropna(subset=['Последнее/нынешнее место работы', 
                                 'Последняя/нынешняя должность'], axis=0)


# В признаке 'Опыт работы (месяц)' заменяем NaN на медианное значение
hh_database['Опыт работы (месяц)'] = \
    hh_database['Опыт работы (месяц)'].fillna(hh_database['Опыт работы (месяц)'].median())


# Определяем среднее по признаку 'Опыт работы (месяц)' после изменений
exp_mean = hh_database['Опыт работы (месяц)'].mean()
print(f'Средний опыт работы в составляет {round(exp_mean)} месяцев')

Средний опыт работы в составляет 114 месяцев


4. Мы добрались до ликвидации выбросов. Сначала очистим данные вручную. Удалите резюме, в которых указана заработная плата либо выше 1 млн. рублей, либо ниже 1 тыс. рублей.

In [38]:
# Выделяем данные между 1тыс. и 1 млн.
mask_1 = hh_database['ЗП (руб)'] > 10**6
mask_2 = hh_database['ЗП (руб)'] < 10**3
outliers = hh_database[mask_1 | mask_2]

# Удаляем выбросы из hh_database
hh_database = hh_database.drop(outliers.index)
print(f'Количество выбросов составило {outliers.shape[0]}')

Количество выбросов составило 89


5. В процессе разведывательного анализа мы обнаружили резюме, в которых **опыт работы в годах превышал возраст соискателя**. Найдите такие резюме и удалите их из данных


In [39]:
# Выделяем данные в которых опыт рабооты больше возраста
mask = hh_database['Опыт работы (месяц)']/12 > hh_database['Возраст']
outliers = hh_database[mask]

# Удаляем выбросы из hh_data
hh_database = hh_database.drop(outliers.index)
print(f'Количество выбросов составило {outliers.shape[0]}')

Количество выбросов составило 7


6. В результате анализа мы обнаружили потенциальные выбросы в признаке **"Возраст"**. Это оказались резюме людей чересчур преклонного возраста для поиска работы. Попробуйте построить распределение признака в **логарифмическом масштабе**. Добавьте к графику линии, отображающие **среднее и границы интервала метода трех сигм**. Напомним, сделать это можно с помощью метода axvline. Например, для построение линии среднего будет иметь вид:

`histplot.axvline(log_age.mean(), color='k', lw=2)`

В какую сторону асимметрично логарифмическое распределение? Напишите об этом в комментарии к графику.
Найдите выбросы с помощью метода z-отклонения и удалите их из данных, используйте логарифмический масштаб. Давайте сделаем послабление на **1 сигму** (возьмите 4 сигмы) в **правую сторону**.

Выведите таблицу с полученными выбросами и оцените, с каким возрастом соискатели попадают под категорию выбросов?

In [40]:
# Создаем признак 'Возраст_log', в котором котором 
# будет записан log-преобразованный признак 'Возраст'
hh_database['Возраст_log'] = np.log(hh_database['Возраст'])


# Создаем гистограмму, отображающую 'Возраст_log'
hist_age = px.histogram(
    data_frame=hh_database,
    x=np.log(hh_database['Возраст']),
    title='Распределение соискателей по возрасту<br>\
(в логарифмическом масштабе)',
    width=500,
    height=500,
    nbins =95
)

# Добавляем линии, отображающие среднее плюс-минус 3 SD
age_log_mean = hh_database['Возраст_log'].mean()
age_log_std = hh_database['Возраст_log'].describe().loc['std']

hist_age.add_vline(x=age_log_mean,  
                   line_color='darkred', 
                   line_width=2.6)

hist_age.add_vline(x=age_log_mean-3*age_log_std,  
                   line_color='black', 
                   line_dash='dash',
                   line_width=2.6)

hist_age.add_vline(x=age_log_mean+3*age_log_std,  
                   line_color='black', 
                   line_dash='dash',
                   line_width=2.6)

hist_age.update_layout(bargap=0.1,
                       xaxis_title='log(Возраст)',
                       yaxis_title='Количество')

hist_age.show()
hist_age.write_html('data/age_log.html')

***Выводы***:
1. Распределение имет смещение в левую сторону.
2. К выбросам относятся два соискателя 15-ти летнего возраста и один соискатель которому 100 лет.